mlcvs/cvs/unsupervised/vae.py

```diff
@@ -0,0 +1,200 @@
+            input descriptors of the CV after preprocessing.
+
+        Returns
+        -------
+        cv : torch.Tensor
+            Shape ``(n_batches, n_cvs)``. The CVs, i.e., the mean output of the
+            encoder (the variance output is discarded).
+        """
+        if self.normIn is not None:
+            x = self.normIn(x)
+        x = self.encoder(x)
+
+        # Take only the means and ignore the log variances.
+        return x[..., :self.n_cvs]
+
+    def encode_decode(self, x: torch.tensor) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        """Run a pass of encoding + decoding.
+
+        The function applies the normalizing to the inputs and its reverse on
+        the output.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            Shape ``(n_batches, n_descriptors)`` or ``(n_descriptors,)``. The
+            input descriptors of the CV after preprocessing.
+
+        Returns
+        -------
+        mean : torch.Tensor
+            Shape ``(n_batches, n_cvs)`` of ``(n_cvs,)``. The mean of the
+            Gaussian distribution associated to the input in latent space.
+        log_variance : torch.Tensor
+            Shape ``(n_batches, n_cvs)`` of ``(n_cvs,)``. The logarithm of the
+            variance of the Gaussian distribution associated to the input in
+            latent space.
+        x_hat : torch.Tensor
+            Shape ``(n_batches, n_descriptors)`` or ``(n_descriptors,)``. The
+            reconstructed descriptors.
+        """
+        # Normalize inputs.
+        if self.normIn is not None:
+            x = self.normIn(x)
+
+        # Encode input into a Gaussian distribution.
+        x = self.encoder(x)
+        mean, log_variance = x[..., :self.n_cvs], x[..., self.n_cvs:]
+
+        # Sample from the Gaussian distribution in latent space.
+        std = torch.exp(log_variance / 2)
+        z = torch.distributions.Normal(mean, std).rsample()
+
+        # Decode sample.
+        x_hat = self.decoder(z)
+        if self.normIn is not None:
+            x_hat = self.normIn.inverse(x)
+
+        return mean, log_variance, x_hat
+
+    def loss_function(self, diff, mean, log_variance, **kwargs):
+        """ELBO loss function when latent space and reconstruction distributions are Gaussians."""
+        return elbo_gaussians_loss(diff, mean, log_variance, **kwargs)
+
+    def training_step(self, train_batch, batch_idx):
+        """Single training step performed by the PyTorch Lightning Trainer."""
+        options = self.loss_options.copy()
+        x = train_batch['data']
+        if 'weights' in train_batch:
+            options['weights'] = train_batch['weights']
+
+        # TODO: Should we do preprocessing here?
+
+        # Encode/decode.
+        mean, log_variance, x_hat = self.encode_decode(x)
+
+        # Loss function.
+        diff = x - x_hat
+        loss = self.loss_function(diff, mean, log_variance, **options)
```

Comment:
do you think it makes sense to make the loss function return also the contributions to the loss (reconstruction and KL) and log them?

Reply:
Yeah, I think that's a good idea. Maybe we could return only the loss as default and control whether the decomposition in kl and reconstruction is returned with an optional argument of ``elbo_gaussians_loss()``?